### PR TITLE
sets modules for tf 0.12

### PR DESCRIPTION
**Pull request by venturel**

### CODE REVIEW

**Review by iuriaranda — CHANGES_REQUESTED** [review]
The upgrade of this module to Terraform 0.12 should include a complete overhaul and simplification. Thanks to tf 0.12 new syntax we don't need to have separate modules for different ELBs configurations: all the `elb_*` modules can be merged into one.

File: elb/variables.tf [hunk 1]
```diff
@@ -42,6 +43,7 @@ variable "connection_draining_timeout" {
 
 variable "access_logs_bucket" {
   description = "The S3 bucket name to store the logs in."
+  default     = ""
```
Comment (iuriaranda) [inline elb/variables.tf]: Default should be `null`, and the same for `access_logs_bucket_prefix`

File: elb/sg.tf [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+resource "aws_security_group_rule" "allow_elb_incoming_from_world" {
+  security_group_id = aws_security_group.elb.id
+  type              = "ingress"
+  from_port         = var.lb_port
+  to_port           = var.lb_port
+  protocol          = "tcp"
+  cidr_blocks       = var.ingoing_allowed_ips
+}
+
+resource "aws_security_group_rule" "allow_elb_incoming_secure_from_world" {
```
Comment (iuriaranda) [inline elb/sg.tf:16]: This is not needed if there's no ssl listener

File: elb/sg.tf [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  security_group_id        = aws_security_group.elb.id
+  type                     = "egress"
+  from_port                = var.instance_port
+  to_port                  = var.instance_port
+  protocol                 = "tcp"
+  source_security_group_id = var.backend_security_groups[count.index]
+}
+
+resource "aws_security_group_rule" "allow_elb_outgoing_secure_to_backend" {
+  count                    = var.instance_ssl_port == var.instance_port ? 0 : var.backend_security_groups_count
```
Comment (iuriaranda) [inline elb/sg.tf]: not needed if there's no ssl listener

File: alb_listener/main.tf [hunk 4]
```diff
@@ -1,42 +1,57 @@
...
-  deregistration_delay = "${var.target_deregistration_delay}"
-  stickiness           = ["${var.target_stickiness}"]
+  port                 = var.target_port
+  protocol             = var.target_protocol
+  vpc_id               = var.vpc_id
+  deregistration_delay = var.target_deregistration_delay
+  dynamic "stickiness" {
+    for_each = [var.target_stickiness]
+    content {
+      # TF-UPGRADE-TODO: The automatic upgrade tool can't predict
```
Comment (iuriaranda) [inline alb_listener/main.tf]: Leftover comment

File: alb_rule_target/main.tf [hunk 5]
```diff
@@ -1,40 +1,55 @@
...
-  stickiness           = ["${var.target_stickiness}"]
-  target_type          = "${var.target_type}"
+  port                 = var.target_port
+  protocol             = var.target_protocol
+  vpc_id               = var.vpc_id
+  deregistration_delay = var.target_deregistration_delay
+  dynamic "stickiness" {
+    for_each = [var.target_stickiness]
+    content {
+      # TF-UPGRADE-TODO: The automatic upgrade tool can't predict
```
Comment (iuriaranda) [inline alb_rule_target/main.tf]: leftover comment

File: alb/main.tf [hunk 6]
```diff
@@ -1,16 +1,26 @@
 # Create a new load balancer
 resource "aws_lb" "alb" {
   load_balancer_type         = "application"
-  name_prefix                = "${var.name_prefix}"
-  internal                   = "${var.internal}"
-  subnets                    = ["${var.subnets}"]
-  security_groups            = ["${aws_security_group.sg_alb.id}"]
-  enable_deletion_protection = "${var.enable_deletion_protection}"
-  access_logs                = ["${var.access_logs}"]
+  name                       = "${var.project}-${var.environment}-${var.name_prefix}"
```
Comment (iuriaranda) [inline alb/main.tf]: This will cause breaking changes at the AWS level and should be carefully thought out. So imo it should not be part of this PR and should be discussed as part of a more general naming strategy to apply to all our modules.

File: README.md [hunk 7]
```diff
@@ -596,14 +311,14 @@ module "elb" {
 
 ```hcl
 module "elb" {
-  source                        = "github.com/skyscrapers/terraform-loadbalancers//elb_with_ssl_with_s3logs"
+  source                        = "github.com/skyscrapers/terraform-loadbalancers//elb"
   name                          = "frontend"
-  subnets                       = ["${module.vpc.frontend_public_subnets}"]
+  subnets                       = [module.vpc.frontend_public_subnets]
```
Comment (iuriaranda) [inline README.md]: No need for a list notation here

File: alb_listener/variables.tf [hunk 8]
```diff
@@ -59,7 +59,7 @@ variable "target_deregistration_delay" {
 
 variable "target_stickiness" {
   description = "List(optional, []): An ALB target_group stickiness block"
-  type        = "list"
+  type        = list(string)
```
Comment (iuriaranda) [inline alb_listener/variables.tf:62]: These type changes should also be reflected in the readme

File: elb/variables.tf [hunk 9]
```diff
@@ -96,7 +103,8 @@ variable "lb_ssl_protocol" {
 }
 
 variable "ssl_certificate_id" {
-  description = " The ARN of an SSL certificate you have uploaded to AWS IAM. Only valid when lb_protocol is either HTTPS or SSL"
+  description = " The ARN of an SSL certificate you have uploaded to AWS IAM. If specified it will create an additional ssl listener"
```
Comment (iuriaranda) [inline elb/variables.tf:106]: This variable is not documented in the readme I think

**Review by iuriaranda — CHANGES_REQUESTED** [review]

File: README.md [hunk 10]
```diff
@@ -12,15 +12,16 @@ Setup an ALB with related resources.
 
 ### Available variables
 
-* [`name_prefix`]: String(required): Name prefix of the ALB and security group
+* [`name_prefix`]: String(optional): Name prefix of the ALB and security group
```
Comment (iuriaranda) [inline README.md:15]: To keep full backwards compatibility and make it more understandable this is how it should work imo:

- set default of `name_prefix` to an empty string
- set default of `name` as `null`

And this is the behaviour

- if none are provided, the name will be completely randomised
- if only `name_prefix` is provided it'll keep the current behaviour
- setting the `name` variable overrides whatever is set in `name_prefix` and sets the "${var.project}-${var.environment}-${var.name}" name style

File: README.md [hunk 11]
```diff
@@ -12,15 +12,16 @@ Setup an ALB with related resources.
...
+* [`name`]: String(optional): Name prefix of the ALB and security group in the format var.project-var.environment-var.name
```
Comment (iuriaranda) [inline README.md]: Technically speaking this is a "name suffix" or "name component". "name prefix" might be a bit misleading. Also for the ELB module.

**Review by iuriaranda — APPROVED** [review]
Just a "would be nice to fix" thing

File: alb_listener/main.tf [hunk 12]
```diff
@@ -1,42 +1,53 @@
...
-  protocol             = "${var.target_protocol}"
-  vpc_id               = "${var.vpc_id}"
-  deregistration_delay = "${var.target_deregistration_delay}"
-  stickiness           = ["${var.target_stickiness}"]
+  port                 = var.target_port
+  protocol             = var.target_protocol
+  vpc_id               = var.vpc_id
+  deregistration_delay = var.target_deregistration_delay
+  dynamic "stickiness" {
+    for_each = [var.target_stickiness]
```
Comment (duboisph) [inline alb_listener/main.tf]: Since it's already a list, don't think you need the `[ ]` brackets

File: alb_rule_target/main.tf [hunk 13]
```diff
@@ -1,40 +1,50 @@
...
-  vpc_id               = "${var.vpc_id}"
-  deregistration_delay = "${var.target_deregistration_delay}"
-  stickiness           = ["${var.target_stickiness}"]
-  target_type          = "${var.target_type}"
+  port                 = var.target_port
+  protocol             = var.target_protocol
+  vpc_id               = var.vpc_id
+  deregistration_delay = var.target_deregistration_delay
+  dynamic "stickiness" {
+    for_each = [var.target_stickiness]
```
Comment (duboisph) [inline alb_rule_target/main.tf]: Same here, you don't need the brackets

**Review by duboisph — APPROVED** [review]
